Content.Shared/Turnstile/Systems/TurnstileSystem.cs

```diff
@@ -0,0 +1,183 @@
+
+        // We have to set collidable to false for one frame, otherwise the client does not
+        // predict the removal of the contacts properly, and instead thinks the mob is colliding when it isn't.
+        SetCollidable((ent, ent.Comp, null), false);
+        SetState((ent, ent.Comp), TurnstileState.Rotating);
+    }
+
+    /// <summary>
+    ///     Turnstiles stop rotating once their admitted entity has passed through.
+    /// </summary>
+    private void StopRotatingIfEntityPassed(Entity<TurnstileComponent> ent)
+    {
+        if (ent.Comp.State != TurnstileState.Rotating)
+            return;
+
+        // Check and see if we're still colliding with the admitted entity.
+        var stillCollidingWithAdmitted = GetIsCollidingWithAdmitted(ent);
+        if (stillCollidingWithAdmitted)
+            return;
+
+        SetState(ent, TurnstileState.Idle);
+        RemComp<PreventCollideComponent>(ent);
+    }
+
+    private bool GetIsCollidingWithAdmitted(EntityUid uid, TurnstileComponent? turnstile = null, TransformComponent? xform = null, PreventCollideComponent? preventCollide = null)
+    {
+        if (!Resolve(uid, ref turnstile) || !Resolve(uid, ref xform) || !Resolve(uid, ref preventCollide))
+            return false;
+
+        if (!TryComp<TransformComponent>(preventCollide.Uid, out var otherxform))
+            return false;
+
+        var turnstileAABB = _entityLookupSystem.GetAABBNoContainer(uid, xform.LocalPosition, xform.LocalRotation);
+        var otherAABB = _entityLookupSystem.GetAABBNoContainer(uid, otherxform.LocalPosition, otherxform.LocalRotation);
+
+        bool isColliding = turnstileAABB.Intersects(otherAABB);
+
+        // Is the entity pulling something else through too? If so, the turnstile should still be spinning to allow this object through.
+        if (TryComp<SharedPullerComponent>(preventCollide.Uid, out var puller))
+        {
+            if (puller.Pulling != null)
+            {
+                if (TryComp<TransformComponent>(puller.Pulling.Value, out otherxform))
+                {
+                    otherAABB = _entityLookupSystem.GetAABBNoContainer(uid, otherxform.LocalPosition, otherxform.LocalRotation);
+                    isColliding |= turnstileAABB.Intersects(otherAABB);
+                }
+            }
+        }
+
+        return isColliding;
+    }
+
+    /// <summary>
+    ///     Iterate over active turnstiles and finish the rotation-admission cycle when mobs leave them.
+    /// </summary>
+    public override void Update(float frameTime)
+    {
+        var query = EntityQueryEnumerator<TurnstileComponent, ActiveTurnstileMarkerComponent>();
+        while (query.MoveNext(out var uid, out var turnstile, out _))
+        {
+            if (Paused(uid))
+                continue;
+
+            SetCollidable(uid, true);
+            StopRotatingIfEntityPassed((uid, turnstile));
+        }
+    }
```

Comment:
Paused is pointless here and toggling collision for a frame is still a code smell, I'd still rather have a joint version using a revolute joint.

Reply:
I'll work at doing just that tomorrow, then. Thank you for your feedback, Sloth!